WordPress/gutenberg
# Post Content Block: Render readonly content as blocks to preserve block supports styles

## Description

This is an exploratory PR to come up with a shorter-term fix for #35376. By rendering the post content block as a live preview within the editor, we resolve the following issues:

* Correctly render the social links icons in this context (resolves: #35376).
* Ensure Duotone filters are rendered for image blocks within post content (resolves part of https://github.com/WordPress/gutenberg/issues/35331).
* Ensure custom link colors are rendered within the post content block.

We achieve the live block preview by doing the following:

* Add a `useDisabled` hook to provide a hook-based approach that achieves most of the functionality of the `Disabled` component, without adding a wrapper element to the DOM.
* Add a `useBlockPreview` hook, that behaves similarly to the `useInnerBlocksProps` hook, by allowing the consuming component to add the behaviour of a block preview to a wrapper element. This turns a component into a block preview, without adding an additional wrapper DOM element.
* The `useBlockPreview` hook uses `useDisabled` internally. The Post Content block is then updated to use `useBlockPreview` for the live block preview.

## To-do

* [x] The appender is rendered within the BlockPreview, but cannot be clicked on (fixed)
* [x] Because the BlockPreview renders the blocks in a Disabled state, the Social Links icons are greyed out (fixed — added a CSS rule)
* [x] Placeholders are rendered (fixed — added a CSS rule) 
* Clicking between two Post Content blocks results in the vertical scroll of the page going to the wrong place — this appears to be behaving similarly on trunk — I believe it's specific to clicking between two similar blocks within a Query loop block within the editor, so might be something for us to look into separately from this PR. Update: I have a fix for the vertical scroll issue in #35945 and have raised the issue with clicking between blocks within the Post Template block in #35944

## How has this been tested?




* Activate an FSE theme like TT1-Blocks
* Publish a post containing the following:
  - An image block using the Duotone filter
  - Social links blocks
  - A block with a custom link color (e.g. Paragraph block, and insert an inline link)
* Open up the site editor, and if a Query loop block does not yet exist in your template, insert the block, and within it, use the Post Content block
* The post you previously published should be rendered correctly in the editor

## Screenshots 

Note the following includes blocks that use Layout spacing, Duotone, and a custom link color (block supports that we need to ensure render correctly within the read only mode of the Post Content block):

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/138647208-7b14f6df-922f-40c9-89fe-cc8c953e62b5.png) | ![image](https://user-images.githubusercontent.com/14988353/138639917-a8f8963d-c4e9-4f68-b9cf-2e301fc38baa.png) |

## Types of changes




Bug fix (non-breaking change which fixes an issue)

## Checklist:
